### PR TITLE
Add function iter_indices, which is a generalization of np.ndindex()

### PR DESCRIPTION
**Pull request by asmeurer**

### CODE REVIEW

**Comment by asmeurer** [comment]
@honno could you help review this? This function should be useful in the array API test suite. It generalises the current `ndindex` helper function in two ways:

- It generates indices for multiple shapes at once. This is useful for elementwise tests where we currently have to broadcast the inputs first.
- I allows skipping axes of arrays. This is useful in the stacking helper in the linalg tests for functions like `cross` and `tensordot` that allow stacking along arbitrary axes.

I plan to either copy this function into the test suite, or to make ndindex a dependency of the test suite. I haven't decided yet which I want to do. If the latter, I will need to make it so that numpy is not a hard dependency of ndindex, so that it doesn't become a hard dependency of the test suite. I still haven't determined how viable this will be.

**Comment by asmeurer** [comment]
Turns out my tests here aren't nearly as good as I thought they were due to https://github.com/HypothesisWorks/hypothesis/issues/3170. I need to try to find a workaround for it and make sure the code doesn't have any hidden bugs. I know there's at least one line in the function which is valid but impossible to hit with the current hypothesis strategy.

File: ndindex/ndindex.py [hunk 1]
```diff
@@ -546,6 +547,169 @@ def broadcast_arrays(self):
         """
         return self
 
+def iter_indices(*shapes, skip_axes=(), _debug=False):
+    """
+    Iterate indices for every element of an arrays of shape `shapes`.
+
+    `shapes` should be tuples of shapes, which are broadcast compatible. Each
```
Comment (honno) [inline ndindex/ndindex.py:554]: Documenting `shapes` being a tuple does make sense technically but its a bit confusing to mention when it comes to star args. My initial thought was that `shapes` was a single arg and not star args.
```suggestion
    `shapes` should be broadcast compatible. Each
```

Comment (asmeurer) [reply ndindex/ndindex.py:554]: It's actually saying it "should be tuples", not "should be a tuple", but I can see how this can be confusing.

**Comment by honno** [comment]
(Just tried this for `test_equals` in the xp test suite, works great! Really does simplify those tests a lot. Now to try using skip axes...)
```python
def test_equal(...):
    ...
    ph.assert_result_shape("equal", (left.shape, right.shape), out.shape)
    ...
    scalar_type = dh.get_scalar_type(promoted_dtype)
    for l_idx, r_idx, o_idx in iter_indices(left.shape, right.shape, out.shape):
        expected = scalar_type(_left[l_idx.raw]) == scalar_type(_right[r_idx.raw])
        assert bool(out[o_idx.raw]) == expected
```

**Comment by honno** [comment]
Very pleased with this util. Played around with the axes stuff and it seems to work as expected! I got caught up re-learning/re-working some things I wrote for #39 (smoke and manip tests), so tomorrow I'll make sure I've covered everything.

File: ndindex/ndindex.py [hunk 2]
```diff
@@ -546,6 +547,164 @@ def broadcast_arrays(self):
...
+
+    if isinstance(skip_axes, int):
+        skip_axes = (skip_axes,)
+    _skip_axes = []
+    for a in skip_axes:
+        try:
+            a = ndindex(a).reduce(ndim).args[0]
+        except IndexError:
+            # Raise the same error as NumPy functions that take axis arguments
+            raise AxisError(f"axis {a} is out of bounds for array of dimension {ndim}")
```
Comment (honno) [inline ndindex/ndindex.py:655]:
```suggestion
            raise AxisError(a, ndim)
```
This follows the `np.AxisError` signature, which will still generate the error message you desire. https://github.com/numpy/numpy/blob/f4353323d0be0c3b3c7d0c752d6c135bc6d36465/numpy/core/_exceptions.py#L124

**Comment by honno** [comment]
Other than the failing test for error message consistency with NumPy, this looks good to me. Your big test makes more sense now and looks pretty useful, though mind I still don't 100% grasp it given I'm not super familiar with the `ndindex` related stuff.  I did run the big test with the new singleton shape behaviour in yesterdays Hypothesis release if that's worth anything. It'd be nice if there was an equivalent, well tested function around like `iter_indices` that we could write a comparison test for...

Also: not sure you knew about this, but today I found out today that Hypothesis ships with a `find` function that can be useful in checking what your strategies can and can't generate, e.g.
```python
>>> from hypothesis import find
>>> find(
...     nps.mutually_broadcastable_shapes(num_shapes=3, max_dims=5),   # strategy
...     lambda S: any(len(s) >= 5 for s in S.input_shapes),            # condition to satisfy
... )
BroadcastableShapes(input_shapes=((), (1, 1, 1, 1, 1), ()), result_shape=(1, 1, 1, 1, 1))
>>> find(
...     nps.mutually_broadcastable_shapes(num_shapes=3, max_dims=5),
...     lambda S: any(len(s) >= 7 for s in S.input_shapes),
... )
NoSuchExample: No examples found of condition lambda S: any(len(s) >= 7 for s in S.input_shapes)
```

**Comment by honno** [comment]
@asmeurer let me know if you'd like me to get this in `array-api-tests`, if you don't want to (or still not sure) whether this should be merged in `ndindex`.

**Comment by asmeurer** [comment]
You can copy it over if you are needing this. If I later decide to make it a dependency we can replace it. 

But be aware that there are still some bugs in this function, which I only found recently due to the hypothesis bug, and haven't yet completely figured out.

**Pull request merged**